test/common/config/metadata_test.cc

```diff
@@ -43,6 +48,79 @@ TEST(MetadataTest, MetadataValuePath) {
             ProtobufWkt::Value::KindCase::KIND_NOT_SET);
 }
 
+class TypedMetadataTest : public testing::Test {
+public:
+  TypedMetadataTest() : registered_factory_(foo_factory_) {}
+
+  struct Foo : public TypedMetadata::Object {
+    Foo(std::string name) : name_(name) {}
+    std::string name_;
+  };
+
+  struct Bar : public TypedMetadata::Object {};
+
+  class FooFactory : public TypedMetadataFactory::TypedMetadataFactory {
+  public:
+    const std::string name() const { return "foo"; }
+    // Throws EnvoyException (conversion failure) if d is empty.
+    std::unique_ptr<const TypedMetadata::Object> parse(const ProtobufWkt::Struct& d) const {
+      if (d.fields().find("name") != d.fields().end()) {
+        return std::make_unique<Foo>(d.fields().at("name").string_value());
+      }
+      throw EnvoyException("Cannot create a Foo when metadata is empty.");
+    }
+  };
+
+protected:
+  FooFactory foo_factory_;
+  Registry::InjectFactory<TypedMetadataFactory> registered_factory_;
+};
+
+TEST_F(TypedMetadataTest, OkTest) {
+  ::envoy::api::v2::core::Metadata metadata;
+  (*metadata.mutable_filter_metadata())[foo_factory_.name()] =
+      MessageUtil::keyValueStruct("name", "foo");
+  TypedMetadataImpl<TypedMetadataFactory> typed(metadata);
+  EXPECT_NE(nullptr, typed.get<Foo>(foo_factory_.name()));
+  EXPECT_EQ("foo", typed.get<Foo>(foo_factory_.name())->name_);
+  // A duck is a duck.
+  EXPECT_EQ(nullptr, typed.get<Bar>(foo_factory_.name()));
+}
+
+TEST_F(TypedMetadataTest, NoMetadataTest) {
+  ::envoy::api::v2::core::Metadata metadata;
```

Comment:
Nit: can you just drop the s`::envoy::api`/`envoy::api`/ from these API ones across the PR?

Reply:
Fixed.